pageserver/src/page_service.rs

```diff
@@ -214,9 +214,23 @@ async fn page_service_conn_main(
     // no write timeout is used, because the kernel is assumed to error writes after some time.
     let mut socket = tokio_io_timeout::TimeoutReader::new(socket);
 
+    let default_timeout_ms = 10 * 1000;
+    let socket_timeout_ms = (|| {
+        fail::fail_point!("simulated-bad-compute-connection", |avg_timeout_ms| {
+            if let Some(avg_timeout_ms) = avg_timeout_ms {
+                let avg = avg_timeout_ms.parse::<i64>().unwrap() as f32;
+                let u = rand::random::<f32>();
+                ((1.0 - u).ln() / (-avg)) as u64
+            } else {
+                default_timeout_ms
+            }
+        });
+        default_timeout_ms
+    })();
+
     // timeout should be lower, but trying out multiple days for
     // <https://github.com/neondatabase/neon/issues/4205>
-    socket.set_timeout(Some(std::time::Duration::from_secs(60 * 60 * 24 * 3)));
+    socket.set_timeout(Some(std::time::Duration::from_millis(socket_timeout_ms)));
```

Comment:
Using a timeout to cause disconnections is highly likely to only exercise disconnections in between requests: the latency to serve a getpage request is in the microseconds, so this test setting a 15ms timeout is probably just going to wait until the client is idle, and then disconnect.

As always, the litmus test for this kind of thing is to point wireshark at it and check what kind of behavior you're really getting, but from reading the code it seems like this isn't going to inject the kind of mid-request disconnects that are most stressful.

Reply:
I think the original version of the code attempting to inject disconnections between a request and a response was on the right track: the only problem was that `process_query` couldn't return a value that caused a disconnect.

Perhaps you could add a `cfg(testing)` variant to `QueryError` that `PostgresBackend::process_message` understands to mean "disconnect now".

Then that, combined with the idle timeout case you've got here, would cover both disconnections between requests (the relatively easy case) and disconnections during requests (the slightly harder request, perhaps).

Reply:
> Perhaps you could add a cfg(testing) variant to QueryError that PostgresBackend::process_message understands to mean "disconnect now".

That's a good idea, not sure why I didn't think of it earlier. Just added this (though I didn't make it `#[cfg(feature = testing)]` because it wasn't compiling and I couldn't really figure out why).